### ISSUE TITLE
Make it possible to backup btrfs root volumes

### ISSUE THREAD

**Issue by kbabioch** [digint/btrbk#121]
This is somehow related to #17. I'm using btrfs for my root volume with a subvolume layout similar to the one deployed by openSUSE. In a simplified version it looks like this:

    [root@antares ~]# btrfs subvolume list /
    ID 257 gen 290098 top level 5 path @
    ID 258 gen 290101 top level 5 path home
    ID 259 gen 290101 top level 5 path var/log

It is then mounted in the following way:

    @ -> /
    home -> /home
    var/log -> /var/log

However btrbk has problems with this layout. The volume specific configuration looks something like this:

    volume /mnt/root
      subvolume @
        target send-receive /mnt/backuphdd1/
      subvolume home
        target send-receive /mnt/backuphdd1/
      subvolume var/log
        target send-receive /mnt/backuphdd1/

However this doesn't work, a dry run outputs the following error messages:

    //@
    !!! Aborted: Failed to fetch subvolume detail: cannot find real path for '//@': No such file or directory

    //home
    !!! Aborted: Not a child subvolume of: /

    //var/log
    !!! Aborted: Not a child subvolume of: /

It can mount the whole thing into another path, e.g. into /mnt/, but to be honest I don't like the idea of my root being mounted twice. Can this somehow be fixed?

**Comment by digint** [comment]
If I get you correctly, it still does not work if you have your btrfs root (id=5) filesystem mounted in `/mnt/root`? This should work with your config above.

For further investigation, can you please provide me with output of:

- `btrbk -l debug dryrun`
- `btrfs subvolume show /`
-  `cat /proc/self/mounts | grep btrfs`

**Comment by kbabioch** [comment]
No, you probably misunderstood me. It *does* work with `/mnt/root`. However I don't like the idea to mount my root filesystem twice. Shouldn't it be possible to back it up with the scheme described above, since this is already mounted?

**Comment by digint** [comment]
btrbk needs a place to put the snapshots in, preferably in the same subvolume as the source. With your setup this is not possible: e.g. snapshots from "/home" would have to be created in "@", which is a different subvolume.

While this should be possible with recent kernels, btrbk needs "subvolume" defined in the config to be a child of "volume" (in terms of btrfs tree, not mount points).

This is by design, and there's good reasons for this. E.g. in your setup, you have "home" reside outside of root ("@"), which makes sense, but why would you ever want the snapshots of "home" inside root?

So yes, there is no way around mounting btrfs root somewhere. Consider doing this in a script, e.g.

```
mount /mnt/root
btrbk run
umount /mnt/root
```

**Comment by kbabioch** [comment]
Hm, ok, so probably I've oversimplified it a little bit. The design I envision should actually look something like this:

    [root@antares ~]# btrfs subvolume list /
    ID 257 gen 290098 top level 5 path @
    ID 258 gen 290101 top level 5 path home
    ID 259 gen 290101 top level 5 path var/log
    ...
    ID 259 gen 290101 top level 5 path .snapshots

This will then be mounted in the following way:

    @ -> /
    home -> /home
    var/log -> /var/log
    ...
   .snapshots -> /mnt/snapshots

Snapshosts of @, home and /var/log should then be put into /mnt/snapshots. After fiddling around with btrbk for a while it seems that this is not possible, while this can be done perfectly fine by hand:

     btrfs subvolume snapshot -r / /mnt/snapshots/@.201612191730
     btrfs subvolume snapshot -r /home /mnt/snapshots/home.201612191730

This way I don't have to mount the "root" subvolume, since it only contains subvolumes that have to mounted for themselves. Personally I do think that btrbk should be more agnostic to the subvolume layout, since there are a lot of different ways to skin the cat and there is no technical reason for this restriction. Or am I missing something here?

**Comment by digint** [comment]
> Personally I do think that btrbk should be more agnostic to the subvolume layout, since there are a lot of different ways to skin the cat and there is no technical reason for this restriction. Or am I missing something here?

No, you are not missing anything, technically this is all possible, and I agree that it would be great if btrbk supports it. There are some reasons why btrbk does not support it:

1. Historical reasons. Creating a snapshot in a different subvolume was not possible with older kernels.

2. A main goal of btrbk is to have simple configurations, so I came up with the volume/subvolume/snapshot_dir semantics, which allows to group things. This made perfectly sense with restrictions of (1), and has proven to be very powerful.

3. Simplicity. Handling mount points is a real pain. If e.g. your `.snapshots` folder would be mounted from a different btrfs filesystem, this would not work. Furthermore things can be mounted multiple times, there is chroot, ...

A lot of the mountpoint "pain" was implemented in recent versions of btrbk, so maybe it would not be that hard to support this. I will keep it in mind and analyse it in more depth for the next big iteration / refactoring of btrbk, but don't expect it to come very soon, if ever. I'm sure it will add some more complexity, and I always try to keep things as simple as possible.

**Comment by HaleTom** [comment]
@kbabioch I've put some thought into the best way (so far) to do this, and came up with: [How do I backup my root filesystem? - suggested FAQ entry](https://github.com/digint/btrbk/issues/130). Your comments would be welcome.

**Comment by sandikata** [comment]
Hello!
So, i tried to use your tool, but on external drive, and it says 
/btr_pool/rootfs
!!! Aborted: Not a child subvolume of: /btr_pool

10 root@GA-990-FXA-UD3-FX8320 /btr_pool # cat /etc/btrbk/btrbk.conf                                                                                                                  :(
snapshot_preserve_min       2d
snapshot_preserve          14d

target_preserve_min        no
target_preserve            20d 10w *m

snapshot_dir               btrbk_snapshots

volume /btr_pool
  subvolume rootfs
    target send-receive    /btr_pool/btr_backup/funtoo

and fstab entry 
/	/btr_pool/rootfs	none	bind	0	0

i just don't have enough space on / to create subvolumes on it. So can i get it work with different partition for that, and just this bind?

**Comment by digint** [comment]
@sandikata : the "subvolume" needs to be on the same file system as the "volume" (btrfs cannot create snapshots on separate file systems). Note that creating snapshots does not require much space, so if you don't preserve the snapshots for too long this will not be an issue.

**Comment by sandikata** [comment]
i have just 4 - 5 GB free space on /

**Comment by digint** [comment]
implemented in #407, this should solve all problems described here.

please add new comments there

**Comment by digint** [comment]
features described here are now included in btrbk-0.32.0, closing issue

**Comment by ledlamp** [comment]
this is so confusing, I just want to back up a remote server with btrfs root to local copy, with `target send-receive`, why does this have to be so complicated, just let me back up from `/`.

```
root@nas:/etc/btrbk# cat btrbk.conf
ssh_identity /root/.ssh/id_rsa
volume ssh://avps.owo69.me:2/
        subvolume *
        target send-receive /mnt/btrfs/avps
root@nas:/etc/btrbk# btrbk dryrun

ERROR: process died unexpectedly (btrbk v0.32.6)
Please contact the author: Axel Burri <axel@tty0.ch>

Stack Trace:
----------------------------------------
invalid url: ssh://avps.owo69.me:2 at /usr/bin/btrbk line 2664.
 at /usr/bin/btrbk line 392.
        main::__ANON__("invalid url: ssh://avps.owo69.me:2 at /usr/bin/btrbk line 2664.\x{a}") called at /usr/bin/btrbk line 2664
        main::vinfo("ssh://avps.owo69.me:2", HASH(0x556e2280ac38)) called at /usr/bin/btrbk line 6148
root@nas:/etc/btrbk#
```
